### DISCUSSION TITLE
flux-tree for heterogeneous tasks?

### DISCUSSION THREAD

**Discussion by andre-merzky** [flux-framework/flux-core#3337] [Q&A]
Hi Flux, 

to increase performance (and specifically throughput) when using a private Flux instance within a job larger allocation, we began to look into flux-tree, which seems to be the recommended way to scale throughput.  @jameshcorbett provided us with interesting insights on how Thesis uses flux-tree, and how the flux tree hierarchy is adapted to optimize task-to-resource mapping -- very nice!

It seem to be the case though that (a) flux-tree is designed to exclusively support *homogeneous* tasks, and (b) it is a one-shot solution (a flux-tree instance cannot be reused -- to run a new batch of tasks, a new flux-tree must be created).

Now it happens to be the case that many of our use cases have require the execution of *heterogeneous* tasks.  Additionally, we do not know the complete set of tasks when execution starts, but instead receive a continuous stream of tasks.  Is flux-tree usable in this context?  Ideally, I would like to define a flux hierarchy, instantiate it with `flux-tree`, and then feed tasks to the root of the tree as they arrive, letting flux take care of distributing them to branches (for load balancing) and leaves (for execution).  Is that possible?

If not, what would you advise to increase task throughput in that setup, in order to ensure scaling behaviour?

Many thanks, Andre.

PS: We can do some opportunistic bulking on the task stream (collect tasks for a certain time and then submit that bunch into a separate flux-tree) -- but the problem remains that the tasks can be extremely heterogeneous in both size and runtime.

**Comment by garlick** [comment]
How far short of expectations do we fall within a single Flux instance?  Is there some simple workload we could run that demonstrates the low throughput, and that we could work to improve?  I know our throughput numbers are not that great at present but the good news is there's plenty of areas that could be optimized.

We may have given you the excuse before that we're working to fill in functionality gaps before doing performance work and that still applies, but at some point in the not too distant future we're going to need to make a performance investment and it would be good to have some test cases.

(Leaving the `flux-tree` discussion to others as I'm sadly clueless!)

**Reply by andre-merzky** [reply]
Oh, I am happy to provide stress test use cases :-D  

The main performance metric for us is tasks/second, as it directly impacts utilization in steady state.  For example, we have been driving our PRTE/PMI-X integration with stressing single core workloads (see [here](https://arxiv.org/pdf/1909.03057.pdf), figures 6 and 8 give some idea of how submission rate impacts utilization).  (alas, PRTE is broken on Summit right now, so we don't reach those numbers anymore).

 We have a number of science workloads to back this stress test setup - but most are more heterogeneous really, so would presumably run into those flux-tree constraints. One example use case where throughput is important: seismic analysis of earthquake traces (Princeton): those guys have an ~infinite number of earthquake traces.  The workflow for analysis of one earthquake has the following task types:

  * n MPI task with 384 ranks, 1 GPU + 1 core each, 10 min per earthquake
  * n * m MPI task with 384 ranks, 1 core each, 10 min
  * n * m * o MPI tasks, 16 ranks, 1 core each, 10 min
  * n * m * o single core tasks, 2 min 

(don't ask me what causes that magic number 384 ... :-P )

The large GPU MPI task is a barrier for all other tasks IIRC, otherwise tasks types are not aligned and different task types will continuously be submitted.  Task runtimes given are approximate and will vary significantly - they are *not* uniform in runtime.  Target scale is all of summit (they have enough data to actually use the full machine).   The large tasks are easy, to keep track of the 16 core and single core tasks (which dominate) is the hard part.

 n: as large as possible; m: several hundred; n: 10 .. 50

If we want to use, say, `n==1, m=100, n=25`, we would pack the ~5k  tasks onto approx. 64 nodes (largest task).  The dominating mix of 16 core and single core task would fill those 64 nodes with approx ~800 tasks.  The fast tasks are 2 min, so we would need to replace the fast tasks with a rate of about 6 tasks/second.  So, n==1 would work with flux :-D  n==2 would double that rate, and we would thus reach flux limits (I see 15/sec in tests) at 128 nodes.  The target size of 4k nodes (or lets make it 1k) would fall short by 1 order of magnitude.

Is this useful for you?

**Reply by dongahn** [reply]
@andre-merzky: @SteVwonder and I discussed a little and we have some questions and suggestions for you. (Will get back to as soon as practical -- too many meeting sorry).

**Reply by grondo** [reply]
Thanks @andre-merzky! This is really helpful IMO. I wonder if we can come up with a useful target benchmark.

>  (I see 15/sec in tests) at 128 nodes.

caveat: I have not digested this entire thread, but 15 jobs/s seems quite low for flux.  I wonder if you are hitting a bottleneck that a simple throughput test using `hostname` doesn't hit. (Currently, flux seems to be limited to about ~30-40 job/s in my testing, though it may be machine dependent)

This is with the simple scheduler (in "single" alloc mode, a known limitation)
```
ƒ(s=128) grondo@fluke8:~/git/flux-core.git$ flux python bulksubmit.py jobs/*.json
bulksubmit: Starting...
bulksubmit: scheduling disabled, reason=Testing
bulksubmit: submitted 5000 jobs in 9.40s. 531.77job/s
bulksubmit: scheduling enabled
bulksubmit: First job finished in about 1.126s
|██████████████████████████████████████████████████████████| 100.0% (35.9 job/s)
bulksubmit: Ran 5000 jobs in 139.1s. 35.9 job/s
```

**Reply by jameshcorbett** [reply]
Andre, are you using Flux's Python bindings? And if so, are you using ``flux.submit()`` or ``flux.submit_async()``? The difference could also explain part of the discrepancy between what you expect @grondo and what @andre-merzky sees. With that modification these are the numbers I got on 1 node of Quartz (it's not on fluke, and I didn't modify the scheduler, so it's not totally comparable):

```
bulksubmit: Starting...
bulksubmit: submitted 1025 jobs in 50.62s. 20.25job/s
bulksubmit: First job finished in about 50.624s
|██████████████████████████████████████████████████████████| 100.0% (20.0 job/s)
bulksubmit: Ran 1025 jobs in 51.2s. 20.0 job/s
```

```
bulksubmit: Starting...
bulksubmit: submitted 5000 jobs in 356.38s. 14.03job/s
bulksubmit: First job finished in about 356.407s
|██████████████████████████████████████████████████████████| 100.0% (13.9 job/s)
bulksubmit: Ran 5000 jobs in 359.2s. 13.9 job/s
```

**Reply by andre-merzky** [reply]
> 15 jobs/s seems quite low for flux
> Andre, are you using Flux's Python bindings?
> ... are you using flux.submit() or flux.submit_async()?

I should have mentioned that I did not, by any means, perform a proper benchmarking nor any kind of optimization!  The 15 jobs/s is approx what I saw on trivial tests submitting single node python jobs.  So I may be well off by whatever factor.

Having said that: yes, that is via the Python bindings, and no, that is not an async submit.  We also have the ability to submit from multiple processes / nodes into the same flux instance.  For other execution backends that can improve throughput significantly, and I should also give that a try with Flux.  And `submit_async()` probably would fit that approach (I am assuming here that it just fires off the submit rpc message and does not wait for it to return)?

Let me set up some proper benchmarking to obtain more reliable numbers with both dummy and real workloads (in case that makes a difference), I'll provide the numbers for that.  Let's see if I manage to use `submit_async()` also for comparison, that's a good point anyway!

**Reply by jameshcorbett** [reply]
I've had issues with this, so just putting it out there---if you're going to use ``submit_async()``, you need to pair that with calls to ``reactor_run()`` in order to fire off any callbacks you might have attached to the submission futures. I think, for instance, that ``flux.job.wait()`` won't recognize that jobs have been submitted if you forget to call ``reactor_run``? (Not sure about that.) There is a good example of the usage of ``submit_async`` [here](https://github.com/flux-framework/flux-workflow-examples/tree/master/async-bulk-job-submit) that can be useful for benchmarking.

**Reply by SteVwonder** [reply]
Yeah, there are two gotchas highlighted in that example that @jameshcorbett linked to:

1) `flux.job.wait()` will only work if you do `submit_async(..., waitable=True)` (but it doesn't require any explicit user-level calls to `reactor_run`, those happen "under the hood" as you are blocking in `flux.job.wait()`)
2) if you do `submit_async(...).then(my_cb_func)`, the `my_cb_func` will only fire if you explicitly call `reactor_run` at the user level

**Comment by jameshcorbett** [comment]
@dongahn @SteVwonder for additional context, Andre told me that sometimes workflows consist of fast-running (~seconds) single-core applications and slow-running (~hours) 300-node applications. A little bit of heterogeneity flux-tree could handle, but if you only had (say) a total of 305 nodes, this sort of workflow seemed to me to basically destroy the usefulness of ``flux-tree``.

Themis also has the sort of problem that Andre describes, but the difference is that Themis makes users determine the flux-tree that Themis will use, and presumably the user knows how heterogeneous their jobs' resource requirements are, so will choose their tree intelligently (that is, if they choose a tree at all, which most users don't). But I also don't know of any Themis use-cases that have jobs with such wildly varying resource requirements.

**Comment by dongahn** [comment]
@andre-merzky:

Thank you for the succinct description. 

I think this problem would be similar to that of the MuMMI workflow. Would this be an area where where you can take a prior knowledge from the users and use it to optimize Flux scheduling?

Let me use your Princeton as an example. Assuming that there are enough numbers of tasks of different types that can be injected, and the users can provide the info on those four distinct task types before starting up the workflow, you can build your flux hierarchy as

S * ( 2 * 384 cores + 384 GPUs) nested flux instances
M * 16 core nested instances

Then, large tasks can be submitted to the first nested instances and small tasks can be submitted to the second nested instances. (In the case of MuMMI they actually schedule the larger type jobs at the top level, which should work equally well for you as well).

Now, how to balance the load within the flux instances of the same type will be the upper level decision, and this can be a co-design between Radical and Flux. I can certainly see a thin flux module can do this but perhaps this should be better handled by the workflow manager for better separations of concerns. 

Nevertheless, since you can "specialize" scheduling per-instance basis, you do want to optimize the small instances with high throughput oriented scheduling. 

Now getting back to your main question, how to build an instance tree. 

`flux-tree` is a convenience interface, so you don't have to be confined by its capabilities. I do think it can be used with minor modification for your case, but my suggestion would be to use lower level interfaces to build your hierarchy without having to be limited by `flux-tree`. Radical is a general purpose tool and it probably makes sense you to bend Flux as much as possible for your needs and I see lower level interfaces make sense. 

We may be able to prototype something using `flux mini batch` or `flux mini alloc` interfaces or python equivalent to build your hierarchy. 

Maybe we should schedule a mini-hackathan to go over some of the design parameters for a high bandwidth communication and then follow through with further on-line discussions? 

Or I'd be also happy to push forward this discussion via this thread.

**Reply by andre-merzky** [reply]
@dongahn 

yes, clustering by task description (size, executable, arguments, ...) would be possible, at least to some extent.  We could write wrappers to hide that tasks get different input files and/or arguments, etc.  

Two problems make this difficult in practice.  The minor one is that we obtain tasks at random points in time, as a continuous stream or in chunks.  An example for this would be Replica-Exchange, where a runtime decision is made on when data are to be exchanged between two tasks, based on some science metric - and we have no control over when that event triggers new  task submissions, nor does that happen in any discernible pattern.  We could do some opportunistic buffering though to collect tasks for sorting I guess.  

The main problem though is that we do not know task runtimes.  Like, we don't know them at all - they are simply not part of our task description.  For some use cases task runtimes tend to be uniform (and in a subset of those the scientists can even estimate them) -- but for others we have very long tail distributions.  Just as a recent example: a Covid docking scan can have 10^6 tasks, with a mean task runtime of ~20 seconds and a max task runtime of >3000 seconds (see [here](https://arxiv.org/pdf/2010.10517.pdf), Table 1).  And it is impossible to know which tasks will be slow and which ones fast - that is a workload level property.  


> We may be able to prototype something using flux mini batch or flux mini alloc interfaces or python equivalent to build your hierarchy.

That sounds interesting!  Let us ponder a bit though how far we will be able to get with tuned flux hierarchies.  I'll bring this up on our devel call next week, so that the rest of the group can spend some brain on this also! :-)

**Reply by dongahn** [reply]
> The main problem though is that we do not know task runtimes. Like, we don't know them at all - they are simply not part of our task description. For some use cases task runtimes tend to be uniform (and in a subset of those the scientists can even estimate them) -- but for others we have very long tail distributions. Just as a recent example: a Covid docking scan can have 10^6 tasks, with a mean task runtime of ~20 seconds and a max task runtime of >3000 seconds (see here, Table 1). And it is impossible to know which tasks will be slow and which ones fast - that is a workload level property.

I see.

One of the things I will likely work on as part of our covid19 workflow is to provide a scheduler interface to monitor the scheduler load. Perhaps, each instance scheduler can provide an RPC that a user or WF manager alike can use to measure the load of the instance scheduler. For example, the pending queue depth/length. 

Maybe the upper WF layer can use this logic to equalize the queue length for some sibling flux instances (e.g,. 16 core instances). Such a technique is often used for cloud bursting albeit in a slightly differently way. So such interface will be required for our HPC+Cloud direction as well.

**Reply by dongahn** [reply]
@andre-merzky: 

BTW, with the current push towards exascale systems (target platforms for ECP), I really do think that exploiting the scheduler parallelism is a way to go, especially with any scientific workflows that have throughput-oriented elements. 

One idea: I don't know what the bootstrapping cost of an Radical instance is and whether each instance can isolate its self from other instances that may share the node, but would it a possibility to instantiate a Radical instance inside a each nested instance (e.g., created with flux mini batch) and let those instances schedule/execute tasks in parallel?

Maybe this way, Radical won't need too much re-architecting to implement generalized multilevel scheduling. Just leverage Flux as its components when it needs to increase the scheduling parallelism.

**Reply by andre-merzky** [reply]
@dongahn :

> One idea: I don't know what the bootstrapping cost of an Radical instance is and whether each instance can isolate its self from other instances that may share the node, but would it a possibility to instantiate a Radical instance inside a each nested instance (e.g., created with flux mini batch) and let those instances schedule/execute tasks in parallel?

You put your finger on the spot!  We think partitioning is indeed the way forward for our stack.  We were just hoping we can offload the work on a partitioned scheduler to you :-D  But really, we are on the same page here.

Just to provide some details on the current state (and this discussion may well provide the push to implement the remaining pieces).  Our stack has 3 bootstrapping stages:

  * bootstrap_0: shell script lands on a batch node, ensures python is set up, loads modules, etc.  This is not cheap.
  * bootstrap_2: python: starts agents, one on the same node, possibly additional ones on compute nodes
  * bootstrap_3: python: agents start components like data stager, scheduler, executor on their nodes (batch nodes or compute nodes)

The numbering is not wrong - we are planning to add bootstrap_1 which splits the job allocation into smaller ones, and each partition would get it's own agent/component tree -- and those partitions could very well host their own flux instances.

Two issues need solving: load balancing over partitions, and handling tasks which are larger than a partition.  The first one we think we know how to do.  The second one may be non-trivial, but we need to be able to run large tasks, so that needs resolving (maybe not immediately though).

So we are discussing this in our group right now, lets see when/how we can progress that.

Some Flux questions in that context:
  - we assume that Flux finds the nodes to manage from `$SLURM_NODELIST` (and equivalent in other batch systems).  Would it be sufficient to change that env in order to get Flux to manage such a partition?  Or can we pass an explicit nodelist to Flux?
 
One way to approach the large-job problem would be to maintain one global flux instance *additionally* to the ones on the partitions.  We would then need to drain some partitions to make sure resources are available.  Assuming we go that route:
  - would two Flux instances conflict when managing the same nodes (we would ensure on a higher level that one is idle when the other runs jobs)?
  - can we advice Flux that, for a specific jobs, it should consider a specific set of nodes (so that we don't have to drain all partitions, but only the ones we need for the large job)?  

PS: the scheduler queue monitoring you mentioned in your previous reply certainly would come in handy here...

Thanks, Andre.

**Reply by dongahn** [reply]
@andre-merzky: 

Great! I have to believe the scheme of Radical is pretty compatible with Flux's nesting support. 

> bootstrap_0: shell script lands on a batch node, ensures python is set up, loads modules, etc. This is not cheap.

For each nested instance, this shell script will be the "batch script" on the subset of resource, that goes into the `flux mini batch` interface and you should be able to go through the rest of bootstrapping phases in a nested instance. (This can even go deeper ... though you probably work with two level nesting for testing.) 

> Two issues need solving: load balancing over partitions, and handling tasks which are larger than a partition. The first one we think we know how to do. T

Yes. And as I mentioned above, I will likely need to add some more scheduler interface to tell its user about the load of the instance. So please keep your suggestions coming if such an interface can help your job better or easier. 

> The second one may be non-trivial, but we need to be able to run large tasks, so that needs resolving (maybe not immediately though).

When a larger job comes into the top level Radical, which cannot scheduler because of other nested instances, you should be able to drain the queue of some of the nested instances (these are just jobs from the perspective of the top level) and then let those nested instances to finish. Once enough nested instances complete, you should be able to scheduler the larger job. 

Some more thoughts/experiments are required, but this is really within the reach of what we can do without bending the proposed scheme too much.

The other thing is to utilize "grow" and "shrink" since we are being headed toward "elastic scheduling" which will be needed for Cloud resources etc. But this requires a lot more research by our researchers. This is the major topic for @milroy.

**Reply by dongahn** [reply]
> we assume that Flux finds the nodes to manage from $SLURM_NODELIST (and equivalent in other batch systems). Would it be sufficient to change that env in order to get Flux to manage such a partition? Or can we pass an explicit nodelist to Flux?

At the top level, we use a single "srun" (or equivalent) to launch an instance (flux start) across all nodes and each broker discover fine-grained resources (cores etc) using hwloc. Once you set up the top level from a foreign resource manager, a nest instance can be created if you submit a nested batch job using flux mini batch. There are other ways to do this but I am just suggesting this as one way with least side effect. The nested instance will again do resource discovery using hwloc (or soon we will do this by using a resource object called RV1 to do this.

**Reply by andre-merzky** [reply]
Great, this all sounds fairly compatible, at least on this abstract level.  I think we have different ideas on what level is doing the partitioning - but having multiple options is good :-)

All right, I am happy with this result of the discussion, thanks a lot for the guidance! -- we'll go back to the drawing board with this information.  We'll do some work with flux-tree anyway -- it seems like the right (simplest) approach for homogeneous workloads, and we have those too, after all :-D

**Reply by dongahn** [reply]
> One way to approach the large-job problem would be to maintain one global flux instance additionally to the ones on the partitions. We would then need to drain some partitions to make sure resources are available. Assuming we go that route:

> - would two Flux instances conflict when managing the same nodes (we would ensure on a higher level that one is idle when the other runs jobs)?
> - can we advice Flux that, for a specific jobs, it should consider a specific set of nodes (so that we don't have to drain all partitions, but only the ones we need for the large job)?

This is an interesting idea as well. I think running two independent flux instances on the overlapping set of resources should be okay. One thing we don't have right now is to tell the scheduler to schedule the job to a specific set of resources (e.g., match by name.) This shouldn't be difficult to add one but we don't have this yet.

**Reply by dongahn** [reply]
> I think we have different ideas on what level is doing the partitioning

Exactly!

But since Flux has worked this issues out for the general case, we might as well exploit that as much as possible and see if this model fits the bill with little or no modifications?

I sort of like the "resource overlapping" instances as well, though. I wonder perhaps this is something that Flux can generally solve so that tools like Radical can leverage it. I have some vague idea but doing this robustly and most effectively will require real use cases and co-design; and it seems your team is bringing in real use cases.

**Reply by andre-merzky** [reply]
> But since Flux has worked this issues out for the general case, we might as well exploit that as much as possible and see if this model fits the bill with little or no modifications?

Yes and no :-)  I certainly see the advantage - but RCT uses partitioning for other reasons also, and last but not least we also have to work with other execution backends than Flux - at least until you achieved world domination :-D

**Reply by dongahn** [reply]
> This is an interesting idea as well. I think running two independent flux instances on the overlapping set of resources should be okay. One thing we don't have right now is to tell the scheduler to schedule the job to a specific set of resources (e.g., match by name.) This shouldn't be difficult to add one but we don't have this yet.

Maybe you won't need to run independent flux instances after all. You should actually be able to submit the large job to the top instance and as you kill the enough smaller flux instances, the large job will ultimately scheduled by the top level. The difference would be some of these smaller instances will not persist so you will have to resubmit them to start after the large job is done.

**Reply by dongahn** [reply]
In case you haven't already noticed, Flux provides a command like `flux queue drain` and `flux queue undrain` which will block until all of the pending jobs will complete. Maybe these commands can come in handy to manage smaller flux instances in lie of large job support.

**Reply by dongahn** [reply]
Finally, we haven't worked out our details yet, but as part of our roadmap, we need to support "partial" free of resources. If this comes,  this may also help increase your performance. Since you don't generally know the actual runtime of your tasks. When you drain the queue for your nested instance, this may lead to significantly resource underutilization. But if the resources that cannot be used by the nested instance  incrementally return to the parent, this should speed up the execution of the larger job.

**Comment by andre-merzky** [comment]
@dongahn : this all sounds indeed very useful and seems to scratch the itches we'll likely face.  It is probably a bit too early to give you more informed feedback on what approach (queue watch, queue drain, grow/shrink, resource reclaim, node-IDs for jobspecs, ...) would be best to focus on, but they all seem to support the dynamism and heterogeneity we target - cool! 

ack on the underutilization on the queue drain - that is only viable if the workflow has a barrier when switching task types (which is the case for some workflows though).

**Comment by dongahn** [comment]
> ack on the underutilization on the queue drain - that is only viable if the workflow has a barrier when switching task types (which is the case for some workflows though).

Just to make sure we are on the same page, I know you know the resource underutilization is the major trade-ff associated with resource partitioning. We are addressing job scheduling throughput with scheduler parallelism using hierarchical scheduling, but resources will be fragmented and this can lead to underutilization. 

I think 3-pronged approach seems  particularly useful:
- Choose the sizes of nested instances judiciously -- if prior knowledge about the task types can be provided to the workflow system, this can be mitigated quite a bit
- When you have to free some nested instances to schedule a large job at the top instance, it would be best to free those who will complete their workloads soonest. But in general, without knowing the task runtime (which is your case), it would be difficult or if not impossible. Here, an ability for a nest instance to free resources (partitiall) which it will no longer be able to use can help. Of elasticity Flux is going after, we will likely get to "shrink" first. So I'm suggesting that as your consideration
- More coarse version of this is to set the time limit for each nested instance as a job. (Heuristics). This may allow the top level instance to find the earliest time a large job can be scheduled.

**Reply by dongahn** [reply]
> When you have to free some nested instances to schedule a large job at the top instance, it would be best to free those who will complete their workloads soonest. But in general, without knowing the task runtime (which is your case), it would be difficult or if not impossible. Here, an ability for a nest instance to free resources (partitiall) which it will no longer be able to use can help. Of elasticity Flux is going after, we will likely get to "shrink" first. So I'm suggesting that as your consideration

https://github.com/flux-framework/rfc/blob/668c2059ccd196845366a260fad6ede7ef42bfd1/spec_28.rst has our next target for Resource Acquisition Protocol Version 1, which an instance scheduler uses to acquire/release resources via its flux-core resource service. Grow and shrink are there to expand later on. We will heavily use this protocol and resource spec specification to support emerging diverse resource types including Cloud resources and multi-tiered storage.
